include/lbann/utils/visitor_hooks.hpp

```diff
@@ -0,0 +1,75 @@
+////////////////////////////////////////////////////////////////////////////////xecu
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_VISITOR_HOOKS_HPP_INCLUDED
+#define LBANN_VISITOR_HOOKS_HPP_INCLUDED
+
+#include "lbann/utils/enum_iterator.hpp"
+
+#include <iostream>
+#include <string>
+
+namespace lbann {
+
+/// Neural network execution mode
+enum class visitor_hook{setup_begin,
+    setup_end,
+    train_begin,
+    train_end,
+    phase_end,
+    epoch_begin,
+    epoch_end,
+    batch_begin,
+    batch_end,
+    test_begin,
+    test_end,
+    validation_begin,
+    validation_end,
```

Comment:
Visitor hooks should be independent of execution modes, so the validation and test epochs should be indistinguishable. If the callback wants to know the execution mode, it should explicitly ask the execution context.
```suggestion
    evaluate_begin,
    evaluate_end,
```

Reply:
Okay @timmoon10, I have updated the visitor hook phases to remove execution mode specific ones and create new `execution_mode_[begin|end]` phases and so on that are combined with the execution mode when they are displayed.